keps/sig-storage/20200217-namespace-transfer.md

```diff
@@ -0,0 +1,283 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP proposes an API to securely transfer Kubernetes storage resources between namespaces.
+The initial implementation will focus on PersistentVolumeClaims.  But the API could be applied
+to other resource types as compelling use cases come up.
+
+`StorageTransferRequest` resources are created in the target namespace.
+`StorageTransferApproval` resources are created in the source namespace.
+When a controller detects matching request/approval resources, a transfer in initiated.
+
+When transferring a PersistentVolumeClaim, the associated PersistentVolume will be updated,
+the source PersistentVolumeClaim will be deleted, and the target PersistentVolumeClaim will be created.
+The transfer simply deals with Kubernetes API resources.
+No data on the physical PersistentVolume is accessed/copied. 
+
+## Motivation
+
+Give Kubernetes users the flexibility to easily and securely share data between namespaces.
+
+### Goals
+
+- Define an API for transferring persistent storage resources between namespaces.
+- The API should be compatible with as many existing storage provisioners as possible.
+
+### Non-Goals
+
+- Transfer of resources other than PersistentVolumeClaims will not be discussed.
+- Implementation details only discussed in the context of influencing the API definition.
+
+## Proposal
+
+`StorageTransferRequest` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: PersistentVolumeClaim
+  targetName: bar #this is optional
+```
+
+`StorageTransferApproval` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferApproval
+metadata:
+  name: approve-foo
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: PersistentVolumeClaim
+  targetNamespace: target-namespace
+```
+
+When matching `StorageTransferRequest` and `StorageTransferApproval` resources are detected, the transfer process is initiated.
+
```

Comment:
One thing I thought might improve this, rather than immediately taking the action on this approval object, would it make sense to leverage the pvc datasource model that we've been using for other things?  So for example what I'm thinking is that to import or accept a transfer the new namespace still issues a new PVC request, but they use a `StorageTransferApproval` object like you have here as the `DataSource`.

So the controller would wait for not only the matching requests, but the final even to kick this off would be a create pvc from that transfer datasource.  Maybe that's more of an implementation detail.

Reply:
Sorry, I often reverse who does the request and who does the accept...  As far as the other points, I see both sides (mine and yours).  Would be curious to see if anybody else has thoughts, I'm open to your ideas.  Just want to make sure usability is good and if there's places to leverage existing work we do so.